### DISCUSSION TITLE
Removing specific symbols from debug section

### DISCUSSION THREAD

**Discussion by tex** [davea42/libdwarf-code#192] [Q&A] [answered]
Hello,

can this library be used to create an application that would remove debug information of some specific symbols from existing elf binary?

Thanks!

**Comment by davea42** [comment, marked as answer]
The simple answer is 'NO'. 

DWARF is as compact as possible (depending on compiler-writer
choices) and  that makes changing content challenging:  Essentially reading it all in to a set of graphs
removing parts,  and writing a new object file.  Just removing stuff, or just setting blobs of bytes  to zeros
will destroy the DWARF involved.  The library cannot store or rewrite such graphs ( such was never part of its design).

One thing is doable though.  In strings in the .debug_str section zeroing
out bytes will not destroy the DWARF, it will still be readable. 

Altering non-null ASCII bytes in any string anywhere in DWARF to be any non-null
ASCII character (possibly the dot (.)) will not destroy the DWARF.

For strings in the .debug_info section ( or .debug_line or others) zeroing out any bytes have
a very very high probability of destroying/corrupting the DWARF, making it unreadable.

**Reply by tex** [reply]
Thank you! I also found that zeroing out the first byte of string representing symbols in .debug_str section I do not want debugger to show appears to work.

**Comment by davea42** [comment]
I wrote "One thing is doable though. In strings in the .debug_str section zeroing
out bytes will not destroy the DWARF, it will still be readable".    To avoid destroying
and UTF8 multi-byte code points I should have said:

"In strings in the .debug_str section zeroing
out ASCII bytes will not destroy the DWARF, it will still be readable"

Zeroing out portions of a utf8 codepoint is probably safe, though of course
attempts to print the string may fail somehow.  If dwarfdump encounters
a damaged UTF8 codepoint when printing a name dwarfdump treats
the entire string as  if it was not UTF8 and prints non-ascii bytes
in a uri-style %nn.   There is a nice source out there that  detects 
invalid UTF8 quickly:   http://bjoern.hoehrmann.de/utf-8/decoder/dfa/
and that is embedded in src/bin/dwarfdump/dd_utf8.c

**Reply by tex** [reply]
Thanks. We want only remove our symbols which we know are ASCII.
